File: question2.py
# DAA - Jackfruit Problem - Question 2 - Solution

def compute_lps(pattern):
    
    # Construct the Longest Prefix Suffix (LPS) array.
    # lps[i] = length of the longest proper prefix which is also a suffix for pattern[0..i]

    m = len(pattern)
    lps = [0] * m

    length = 0  # length of previous longest prefix suffix
    i = 1

    while i < m:
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        else:
            if length != 0:
                length = lps[length - 1]
            else:
                lps[i] = 0
                i += 1

    return lps


def kmp_search(text, pattern):

    # Perform KMP pattern matching.
    # Returns list of starting indices where pattern is found in text.

    n = len(text)
    m = len(pattern)

    if m == 0:
        return []

    lps = compute_lps(pattern)

    i = 0  # index for text
    j = 0  # index for pattern

    result = []

    while i < n:
        if text[i] == pattern[j]:
            i += 1
            j += 1

        if j == m:
            result.append(i - j)
            j = lps[j - 1]

        elif i < n and text[i] != pattern[j]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1

    return result


def brute_force_search(text, pattern):
    
    # Brute-force string matching.
    # Returns list of starting indices where pattern is found in text.

    n = len(text)
    m = len(pattern)

    result = []

    for i in range(n - m + 1):
        match = True
        for j in range(m):
            if text[i + j] != pattern[j]:
                match = False
                break
        if match:
            result.append(i)

    return result


if __name__ == "__main__":
    text = "ABABDABACDABABCABAB"
    pattern = "ABABCABAB"

    print("Text:", text)
    print("Pattern:", pattern)

    # Compute LPS
    lps = compute_lps(pattern)
    print("\nLPS Array:", lps)

    # KMP Search
    kmp_result = kmp_search(text, pattern)
    print("\nKMP Matches at indices:", kmp_result)

    # Brute Force Search
    brute_result = brute_force_search(text, pattern)
    print("Brute Force Matches at indices:", brute_result)

File: question2.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Compute LPS (Longest Prefix Suffix) array
void compute_lps(const char *pattern, int m, int *lps)
{
    int length = 0;
    lps[0] = 0;

    int i = 1;
    while (i < m)
    {
        if (pattern[i] == pattern[length])
        {
            length++;
            lps[i] = length;
            i++;
        }
        else
        {
            if (length != 0)
            {
                length = lps[length - 1];
            }
            else
            {
                lps[i] = 0;
                i++;
            }
        }
    }
}

// KMP Search
void kmp_search(const char *text, const char *pattern)
{
    int n = strlen(text);
    int m = strlen(pattern);

    if (m == 0)
        return;

    int *lps = (int *)malloc(m * sizeof(int));
    compute_lps(pattern, m, lps);

    int i = 0; // index for text
    int j = 0; // index for pattern

    printf("\nKMP Matches at indices: ");

    while (i < n)
    {
        if (text[i] == pattern[j])
        {
            i++;
            j++;
        }

        if (j == m)
        {
            printf("%d ", i - j);
            j = lps[j - 1];
        }
        else if (i < n && text[i] != pattern[j])
        {
            if (j != 0)
            {
                j = lps[j - 1];
            }
            else
            {
                i++;
            }
        }
    }

    free(lps);
}

// Brute-force search
void brute_force_search(const char *text, const char *pattern)
{
    int n = strlen(text);
    int m = strlen(pattern);

    printf("\nBrute Force Matches at indices: ");

    for (int i = 0; i <= n - m; i++)
    {
        int match = 1;

        for (int j = 0; j < m; j++)
        {
            if (text[i + j] != pattern[j])
            {
                match = 0;
                break;
            }
        }

        if (match)
        {
            printf("%d ", i);
        }
    }
}

int main()
{
    char text[] = "ABABDABACDABABCABAB";
    char pattern[] = "ABABCABAB";

    int m = strlen(pattern);
    int *lps = (int *)malloc(m * sizeof(int));

    printf("Text: %s\n", text);
    printf("Pattern: %s\n", pattern);

    // Compute LPS
    compute_lps(pattern, m, lps);

    printf("\nLPS Array: ");
    for (int i = 0; i < m; i++)
    {
        printf("%d ", lps[i]);
    }
    printf("\n");

    // KMP Search
    kmp_search(text, pattern);

    // Brute Force Search
    brute_force_search(text, pattern);

    free(lps);
    return 0;
}